lib/active_merchant/billing/gateways/adyen.rb

```diff
@@ -622,20 +622,31 @@ def add_mpi_data_for_network_tokenization_card(post, payment, options)
 
         post[:mpiData] = {}
         post[:mpiData][:authenticationResponse] = 'Y'
-        post[:mpiData][:cavv] = payment.payment_cryptogram
+        if NETWORK_TOKENIZATION_CARD_SOURCE[payment.source.to_s].nil? && options[:switch_cryptogram_mapping_nt]
+          post[:mpiData][:tokenAuthenticationVerificationValue] = payment.payment_cryptogram
+        else
+          post[:mpiData][:cavv] = payment.payment_cryptogram
+        end
         post[:mpiData][:directoryResponse] = 'Y'
         post[:mpiData][:eci] = payment.eci || '07'
       end
 
-      def add_recurring_contract(post, options = {})
-        return unless options[:recurring_contract_type]
+      def add_recurring_contract(post, options = {}, payment = nil)
+        return unless options[:recurring_contract_type] || (payment.try(:source) && options[:switch_cryptogram_mapping_nt])
 
-        post[:recurring] = {}
-        post[:recurring][:contract] = options[:recurring_contract_type]
+        post[:recurring] ||= {}
+        post[:recurring][:contract] = options[:recurring_contract_type] if options[:recurring_contract_type]
         post[:recurring][:recurringDetailName] = options[:recurring_detail_name] if options[:recurring_detail_name]
         post[:recurring][:recurringExpiry] = options[:recurring_expiry] if options[:recurring_expiry]
         post[:recurring][:recurringFrequency] = options[:recurring_frequency] if options[:recurring_frequency]
         post[:recurring][:tokenService] = options[:token_service] if options[:token_service]
+        if payment.try(:source) && options[:switch_cryptogram_mapping_nt]
```

Comment:
The payment will have a source even if it isn't apple_pay or google_pay. We still need to check that it isn't those types just like on line 625 above.
```suggestion
        if NETWORK_TOKENIZATION_CARD_SOURCE[payment.source.to_s].nil? && options[:switch_cryptogram_mapping_nt]
```